### PR TITLE
Fix handling of "infinite" clock offsets.

### PR DESCRIPTION
**Pull request by petermattis**
Infinite clock offsets are generated when a heartbeat times out (6s) or
takes too long to finish (greater than 1.5s). Previously, such offsets
were treated as valid values as part of the cluster offset interval
calculation, but doing so meant that a sufficient number of slow nodes
in a cluster (or a sufficiently slow network) would guarantee an
unhealthy clock offset interval to be calculated.

Now we never add "infinite" offsets to the clock offset monitor map. The
monitor is more pessimistic and will only suicide if the set of
good (non-infinite) clock offsets it has computes a cluster offset
interval that is larger than the max offset interval. If a node cannot
talk to any other nodes then no its cluster offset interval will be
considered healthy.

### CODE REVIEW

**Comment by tamird** [comment]
Can you re-read the commit message? There are some grammatical issues that make it hard to follow

File: rpc/client.go [hunk 1]
```diff
@@ -265,18 +265,19 @@ func (c *Client) heartbeat() error {
 				remoteTimeNow := response.ServerTime + (receiveTime-sendTime)/2
 				c.offset.Offset = remoteTimeNow - receiveTime
 			}
+			offset := c.offset
```
Comment (tamird) [inline rpc/client.go]: these changes are unrelated to the description in the commit message - could they be in a separate commit?

File: rpc/client.go [hunk 2]
```diff
@@ -265,18 +265,19 @@ func (c *Client) heartbeat() error {
...
-		// Allowed twice gossip interval.
+		// Allowed twice heartbeat interval.
 		c.mu.Lock()
 		c.healthy = false
 		c.offset = proto.InfiniteOffset
-		c.offset.MeasuredAt = c.clock.PhysicalNow()
+		offset := c.offset
 		c.mu.Unlock()
-		c.remoteClocks.UpdateOffset(c.addr.String(), c.offset)
+		c.remoteClocks.UpdateOffset(c.addr.String(), offset)
```
Comment (bdarnell) [inline rpc/client.go]: UpdateOffset will always be a no-op here because we set c.offset to InfiniteOffset. Can we just remove everything but the logging from this `case`?

File: rpc/client.go [hunk 3]
```diff
@@ -265,18 +265,19 @@ func (c *Client) heartbeat() error {
 				remoteTimeNow := response.ServerTime + (receiveTime-sendTime)/2
 				c.offset.Offset = remoteTimeNow - receiveTime
 			}
+			offset := c.offset
 			c.mu.Unlock()
-			c.remoteClocks.UpdateOffset(c.addr.String(), c.offset)
+			c.remoteClocks.UpdateOffset(c.addr.String(), offset)
```
Comment (bdarnell) [inline rpc/client.go]: Can we rearrange this so that we only call UpdateOffset for valid (finite) offsets? It looks like we could probably remove proto.InfiniteOffset entirely.

File: rpc/client.go (hunk 1 above)
Comment (petermattis) [inline rpc/client.go]: They could be a separate commit, but its fairly common to fix associated problems in code that is being edited. In this case, I happened to notice that `c.offset` was being used outside of the lock which is protecting it.

File: rpc/client.go (hunk 2 above)
Comment (petermattis) [inline rpc/client.go]: I've changed this up to remove `proto.InfiniteOffset` and not call `UpdateOffset` in this code path. Note that `c.offset` is passed to the server in the heartbeat request.

File: rpc/client.go (hunk 3 above)
Comment (petermattis) [inline rpc/client.go]: Yep. Done.

File: proto/heartbeat.proto [hunk 4]
```diff
@@ -24,11 +24,12 @@ import "gogoproto/gogo.proto";
 
 // RemoteOffset keeps track of this client's estimate of its offset from a
 // remote server. Uncertainty is the maximum error in the reading of this
-// offset, so that the real offset should be in the interval
-// [Offset - Uncertainty, Offset + Uncertainty]. If the last heartbeat timed
-// out, Offset = InfiniteOffset. Offset and Uncertainty are measured using the
-// remote clock reading technique described in
-// http://se.inf.tu-dresden.de/pubs/papers/SRDS1994.pdf, page 6.
+// offset, so that the real offset should be in the interval [Offset -
```
Comment (tamird) [inline proto/heartbeat.proto]: intentional?

Comment (petermattis) [inline proto/heartbeat.proto]: Yes. Was your change to merge the two paragraphs into one intentional?

Comment (tamird) [inline proto/heartbeat.proto]: Oh, it wasn't, but the flowing of `[Offset - Uncertainty, ...]` on one line was

Comment (petermattis) [inline proto/heartbeat.proto]: Fixed.

File: proto/heartbeat.proto [hunk 5]
```diff
@@ -26,9 +26,10 @@ import "gogoproto/gogo.proto";
 // remote server. Uncertainty is the maximum error in the reading of this
 // offset, so that the real offset should be in the interval
 // [Offset - Uncertainty, Offset + Uncertainty]. If the last heartbeat timed
-// out, Offset = InfiniteOffset. Offset and Uncertainty are measured using the
-// remote clock reading technique described in
-// http://se.inf.tu-dresden.de/pubs/papers/SRDS1994.pdf, page 6.
+// out, Offset  = 0.
```
Comment (tamird) [inline proto/heartbeat.proto]: there are two spaces before this `=`

Comment (petermattis) [inline proto/heartbeat.proto]: Yeah, yeah, yeah.

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
LG
